ipc/nodepart/arwenDriver.go

```diff
@@ -0,0 +1,308 @@
+	}
+
+	arwenPath = path.Join(cwd, "arwen")
+	if fileExists(arwenPath) {
+		return arwenPath, nil
+	}
+
+	return "", common.ErrArwenNotFound
+}
+
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+
+	return !info.IsDir()
+}
+
+func (driver *ArwenDriver) resetPipeStreams() error {
+	closeFile(driver.arwenInputRead)
+	closeFile(driver.arwenInputWrite)
+	closeFile(driver.arwenOutputRead)
+	closeFile(driver.arwenOutputWrite)
+	closeFile(driver.arwenLogRead)
+	closeFile(driver.arwenLogWrite)
+
+	var err error
+
+	driver.arwenInputRead, driver.arwenInputWrite, err = os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	driver.arwenOutputRead, driver.arwenOutputWrite, err = os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	driver.arwenLogRead, driver.arwenLogWrite, err = os.Pipe()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func closeFile(file *os.File) {
+	if file != nil {
+		err := file.Close()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot close file.\n")
+		}
+	}
+}
+
+func (driver *ArwenDriver) restartArwenIfNecessary() error {
+	if !driver.IsClosed() {
+		return nil
+	}
+
+	err := driver.startArwen()
+	return err
+}
+
+// IsClosed checks whether the Arwen process is closed
+func (driver *ArwenDriver) IsClosed() bool {
+	pid := driver.command.Process.Pid
+	process, err := os.FindProcess(pid)
+	if err != nil {
+		return true
+	}
+
+	err = process.Signal(syscall.Signal(0))
+	if err != nil {
+		return true
+	}
+
+	return false
+}
+
+// RunSmartContractCreate sends a deploy request to Arwen and waits for the output
+func (driver *ArwenDriver) RunSmartContractCreate(input *vmcommon.ContractCreateInput) (*vmcommon.VMOutput, error) {
+	driver.nodeLogger.Trace("RunSmartContractCreate")
+
+	err := driver.restartArwenIfNecessary()
+	if err != nil {
+		return nil, common.WrapCriticalError(err)
+	}
+
+	request := common.NewMessageContractDeployRequest(input)
+	response, err := driver.part.StartLoop(request)
+	if err != nil {
+		driver.stopArwen()
+		return nil, common.WrapCriticalError(err)
+	}
+
+	typedResponse := response.(*common.MessageContractResponse)
+	return typedResponse.VMOutput, response.GetError()
+}
+
+// RunSmartContractCall sends an execution request to Arwen and waits for the output
+func (driver *ArwenDriver) RunSmartContractCall(input *vmcommon.ContractCallInput) (*vmcommon.VMOutput, error) {
+	driver.nodeLogger.Trace("RunSmartContractCall", "sc", input.RecipientAddr)
+
+	err := driver.restartArwenIfNecessary()
+	if err != nil {
+		return nil, common.WrapCriticalError(err)
+	}
+
+	request := common.NewMessageContractCallRequest(input)
+	response, err := driver.part.StartLoop(request)
+	if err != nil {
+		driver.nodeLogger.Error("RunSmartContractCall", "err", err)
+		driver.stopArwen()
+		return nil, common.WrapCriticalError(err)
+	}
+
+	typedResponse := response.(*common.MessageContractResponse)
+	return typedResponse.VMOutput, response.GetError()
```

Comment:
smart contract processor does not care about result when there is an error.

Reply:
Fixed.